src/transport/http2.rs

```diff
@@ -115,83 +117,47 @@ impl<In: RpcMessage, Out: RpcMessage> Clone for ClientChannel<In, Out> {
-                                    match req_tx.send_async(msg).await {
-                                        Ok(()) => {}
-                                        Err(_cause) => {
-                                            // channel closed
-                                            break;
-                                        }
-                                    }
-                                }
-                                Err(_cause) => {
-                                    // deserialize error
-                                    break;
-                                }
-                            },
-                            Err(_cause) => {
-                                // error from the network layer
-                                break;
-                            }
-                        }
-                    }
-                };
-                let body = res_rx.into_stream().map(|out| {
-                    let data = bincode::serialize(&out)?;
-                    // todo: check size
-                    Ok::<Vec<u8>, bincode::Error>(data)
-                });
-                tokio::spawn(forwarder);
-                // let body = TakeUntilEither::new(body, forwarder);
-                // Create a response with the response body channel as the response body
-                let response = Response::builder()
-                    .status(StatusCode::OK)
-                    .body(Body::wrap_stream(body))
-                    .map_err(|_e| "unable to set body")?;
-
-                Ok::<Response<Body>, String>(response)
-            }
-        };
-        // todo: can this stack of weird stuff be simplified?
+
+        // The hyper "MakeService" which is called for each connection that is made to the
+        // server.  It creates another Service which handles a single request.
         let service = make_service_fn(move |socket: &AddrStream| {
             let remote_addr = socket.remote_addr();
             log!("connection from {:?}", remote_addr);
-            let server_fn = server_fn.clone();
+
+            // Need a new accept_tx to move to the future on every call of this FnMut.
+            let accept_tx = accept_tx.clone();
             async move {
-                let server_fn = server_fn.clone();
-                Ok::<_, Infallible>(service_fn(move |req: Request<Body>| {
-                    let server_fn = server_fn.clone();
-                    server_fn(req)
-                }))
+                let one_req_service = service_fn(move |req: Request<Body>| {
+                    // This closure is an FnMut as well, so clone accept_tx once more.
+                    Self::handle_one_http2_request(req, accept_tx.clone())
+                });
+                Ok::<_, Infallible>(one_req_service)
             }
         });
+
         let mut addr_incomping = AddrIncoming::bind(addr)?;
         addr_incomping.set_nodelay(true);
```

Comment:
I found the nodelay to make almost no difference. Should we remove it?

Reply:
sure, but that's an important enough change to go in it's own PR.  I will follow up.